src/librustc/hir/lowering.rs

```diff
@@ -2154,9 +2155,10 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         } else {
             match decl.output {
                 FunctionRetTy::Ty(ref ty) => match in_band_ty_params {
-                    Some((def_id, _)) if impl_trait_return_allow => {
-                        hir::Return(self.lower_ty(ty, ImplTraitContext::OpaqueTy(Some(def_id))))
-                    }
+                    Some((def_id, _)) if impl_trait_return_allow => hir::Return(self.lower_ty(
+                        ty,
+                        ImplTraitContext::OpaqueTy(Some(def_id), hir::OpaqueTyOrigin::FnReturn),
+                    )),
                     _ => hir::Return(self.lower_ty(ty, ImplTraitContext::disallowed())),
```

Comment:
Consider extracting `hir::Return(self.lower_ty(ty, ctxt))` as a common part of both branches.